packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart

```diff
@@ -181,62 +172,56 @@ class AmplifyPushNotifications extends PushNotificationsPluginInterface {
     await _hostApi.setBadgeCount(badgeCount);
   }
 
+  Future<void> _registerDeviceWhenConfigure() async {
+    late String deviceToken;
+
+    try {
+      deviceToken = await onTokenReceived.first;
+      await _registerDevice(deviceToken);
+    } on Exception catch (error) {
+      // the error mostly like is the App doesn't have corresponding
+      // capability to request a push notification device token
+      throw PushNotificationException(
+        'Error occurred awaiting for device token to register device with Pinpoint',
+        recoverySuggestion: 'Please review the underlying exception',
+        underlyingException: error,
+      );
+    }
+  }
+
   void _foregroundNotificationListener(
     PushNotificationMessage pushNotificationMessage,
+  ) =>
+      _serviceProviderClient.recordNotificationEvent(
+        eventType: PinpointEventType.foregroundMessageReceived,
+        notification: pushNotificationMessage,
+      );
+
+  void _backgroundNotificationListener(
+    PushNotificationMessage pushNotificationMessage,
   ) {
     _serviceProviderClient.recordNotificationEvent(
-      eventType: PinpointEventType.foregroundMessageReceived,
+      eventType: PinpointEventType.backgroundMessageReceived,
       notification: pushNotificationMessage,
     );
   }
 
   void _notificationOpenedListener(
     PushNotificationMessage pushNotificationMessage,
-  ) {
-    _serviceProviderClient.recordNotificationEvent(
-      eventType: PinpointEventType.notificationOpened,
-      notification: pushNotificationMessage,
-    );
-  }
+  ) =>
+      _serviceProviderClient.recordNotificationEvent(
+        eventType: PinpointEventType.notificationOpened,
+        notification: pushNotificationMessage,
+      );
 
-  void _tokenReceivedListener(String deviceToken) {
-    unawaited(_registerDevice(deviceToken));
+  Future<void> _tokenReceivedListener(String deviceToken) async {
+    await _registerDevice(deviceToken);
```

Comment:
`_registerDevice` may throw below, when that happens, the exception may become unhandled exception at this point.